core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotationUtil.java

```diff
@@ -404,26 +443,74 @@ private static void disallowedListAnnotation(Class<? extends Annotation> annotat
      */
     private static List<AnnotatedElement> resolveTargetElements(AnnotatedElement element,
                                                                 EnumSet<FindOption> findOptions) {
-        final List<AnnotatedElement> elements;
-        if (findOptions.contains(FindOption.LOOKUP_SUPER_CLASSES) && element instanceof Class) {
-            final Class<?> superclass = ((Class<?>) element).getSuperclass();
-            if ((superclass == null || superclass == Object.class) &&
-                ((Class<?>) element).getInterfaces().length == 0) {
-                elements = ImmutableList.of(element);
-            } else {
-                final Builder<AnnotatedElement> collector = new Builder<>();
-                collectSuperClasses((Class<?>) element, collector,
-                                    findOptions.contains(FindOption.COLLECT_SUPER_CLASSES_FIRST));
-                elements = collector.build();
-            }
+        if (!findOptions.contains(FindOption.LOOKUP_SUPER_CLASSES)) {
+            return ImmutableList.of(element);
+        }
+
+        final boolean collectSuperClassFirst = findOptions.contains(FindOption.COLLECT_SUPER_CLASSES_FIRST);
+
+        if (element instanceof Class) {
+            final List<Class<?>> classes = collectClasses((Class<?>) element, collectSuperClassFirst);
+            return convertToAnnotatedElements(classes);
+
+        } else if (element instanceof Method && findOptions.contains(FindOption.LOOKUP_SUPER_METHODS)) {
+            final Method method = (Method) element;
```

Comment:
nit: let's use early return in this method:
```suggestion
            return convertToAnnotatedElements(classes);
        }
        if (element instanceof Method && findOptions.contains(FindOption.LOOKUP_SUPER_METHODS)) {
            final Method method = (Method) element;
```